Product Collection - Add `Created` filter in inspector controls

imanish003 opened the pull request:
<!-- Please do not remove any information from this pull request. Instead, add N/A or leave blank if not applicable -->

## What

Users can now easily limit their search results to products created within specific time frames. Here is how the UI of new filter looks like: 
![image](https://github.com/woocommerce/woocommerce-blocks/assets/16707866/1c29049d-1408-4df0-8b7a-be039aa39787)


The code changes include:
- A new 'timeFrame' property added to the DEFAULT_QUERY constant in constants.ts, initialized as null, allowing for the storage of time frame data.
- Creation of a new component `CreatedControl` in created-control.tsx that provides UI elements for selecting a time frame filter.
- Inclusion of `CreatedControl` in the Product Collection Inspector Controls.
- Expansion of the ProductCollectionQuery interface in types.ts to include a 'timeFrame' attribute.
- Addition of the 'timeFrame' parameter handling within the ProductCollection PHP class to construct and execute the date query based on the provided time frame.

Fixes #11357 

## Why

<!-- Describe the reason for your changes. This will help the reviewer and future readers get additional context -->

The motivation behind these changes is to provide store owners with more flexibility and precision when displaying products. By filtering products based on their creation date, store owners can create dynamic collections that highlight new products or inventory changes over specific periods.

## Testing Instructions

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

_Please consider any edge cases this change may have, and also other areas of the product this may impact._

**Case 1: Normal**

1. Create a new post & add "Product Collection (Beta)" block to it.
2. In the inspector sidebar, click on the three dots located within the filters panel.
3. Choose "Created" to add the Created filter to the Inspector sidebar.
4. Adjust the filter settings as follows:
    - **Operator**: Select either "Within" or "Before".
    - **Value**: Choose a predefined range such as "last 24 hours" or "last 7 days."
5. Confirm that the products are filtered according to the criteria you've set. To modify product creation dates for testing purposes:
    - Navigate to **Products** > **All Products**.
    - Select the product you wish to edit.
    - Update the **Published on** date and time to your chosen parameters.
    - Click **Update** to save changes.
6. Publish the post and then review Frontend to ensure that the filter works correctly—only the products within the specified timeframe should be visible.
7. It's crucial to test various operator and value combinations to ensure robust functionality.

Demo Video:

https://github.com/woocommerce/woocommerce-blocks/assets/16707866/315aca24-13a0-464a-8282-00cafee1d908


**Case 2: Resetting the value of the** Created **filter**

1. Refer to the procedure outlined in Case 1 to add *Created* filter into the Inspector sidebar.
2. Modify the operator and value parameters as required.
3. Confirm the functionality of the reset options for the filter using the methods illustrated in the accompanying screenshot:
    1. By selecting the “Created” label.
    2. Alternatively, by pressing the “Reset all” button.
<img src="https://github.com/woocommerce/woocommerce-blocks/assets/16707866/fbfe7d31-bba6-4985-a0ed-2dae89df9b62" width="300">


**Case 3: Ensuring older blocks aren't broken (To be tested by developers only)**

1. Switch to the trunk branch.
2. Create a new post and add "Product Collection (Beta)" block.
3. Save and publish the post.
4. Switch to the branch of this PR i.e., `add/11357-product-collection-filters-creation-date`
5. Verify that everything still functions as expected in both the Editor and Frontend.

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested (ie: it makes changes to tests, coding standards, docblocks, etc.). -->
* [ ] Should be tested by the development team exclusively <!-- Check this box if this PR should be tested by the development team exclusively (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

## Screenshots or screencast

<!-- Any screenshots of UI changes will be helpful to include here. Leave blank if not applicable. -->

| Before | After |
| ------ | ----- |
|        |       |

## WooCommerce Visibility

<!-- Check this documentation link (../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WooCommerce core, part of the feature plugin, or experimental. -->
Required:

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental
* [ ] N/A

## Checklist

Required:
* [x] This PR has either a `[type]` label or a `[skip-changelog]` label.
* [ ] This PR is assigned to a milestone.

Conditional:
* [x] This PR has a changelog description (if `[skip-changelog]` label is not present).
* [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR adds/removes an experimental interfaces, and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR has been accessibility tested.
* [ ] This PR has had any necessary documentation added/updated.

## Changelog
<!-- Provide a brief, descriptive summary of the changes in this PR. Include potential impacts on different parts of the product. Example: "Updated the checkout process to streamline the experience for users and reduce the number of steps." -->

> Product Collection - Add new _Created_ filter in inspector controls

imanish003 committed:
Add time frame filter to Product Collection block

This commit introduces the ability to filter products within the Product Collection block by a specified time frame. The changes include:

- A new 'timeFrame' property added to the DEFAULT_QUERY constant in constants.ts, initialized as null, allowing for the storage of time frame data.
- Creation of a new component `CreatedControl` in created-control.tsx that provides UI elements for selecting a time frame filter.
- Inclusion of `CreatedControl` in the Product Collection Inspector Controls.
- Expansion of the ProductCollectionQuery interface in types.ts to include a 'timeFrame' attribute.
- Addition of the 'timeFrame' parameter handling within the ProductCollection PHP class to construct and execute the date query based on the provided time frame.

The addition of the time frame filter offers enhanced flexibility in presenting products and allows users to dynamically segment their product lists based on product creation dates.

imanish003 committed:
Refactor: Standardize 'timeFrame' to be 'undefined' instead of 'null'

This commit includes a refactoring that changes the initialization and reset values for the `timeFrame` property from `null` to `undefined`. This standardization affects the constants, type definitions, and the handling of the `timeFrame` property in both the inspector controls and the PHP backend.

imanish003 committed:
Merge branch 'trunk' into add/11357-product-collection-filters-creation-date

Aljullu commented on assets/js/blocks/product-collection/inspector-controls/created-control.tsx:
```diff
@@ -0,0 +1,109 @@
+						} }
+						value={ timeFrame?.operator || ETimeFrameOperator.IN }
+					>
+						<ToggleGroupControlOption
+							value={ ETimeFrameOperator.IN }
+							label={ __( 'IN', 'woo-gutenberg-products-block' ) }
+						/>
+						<ToggleGroupControlOption
+							value={ ETimeFrameOperator.NOT_IN }
+							label={ __(
+								'Not in',
+								'woo-gutenberg-products-block'
+							) }
+						/>
```
Nit: is it intentional that `IN` is uppercase while `Not in` isn't?

![imatge](https://github.com/woocommerce/woocommerce-blocks/assets/3616980/88ecc9cd-eba5-4c2e-b99d-23bfe038c327)

imanish003 replied:
@Aljullu, I've referenced the design from the provided screenshot; however, I'm unable to locate it in Figma. @kmanijak, could you please share the Figma design's URL? 🙂

To ensure consistency, we could:
- Capitalize all letters: IN & NOT IN
- Or capitalize only the first letter: In & Not in

@manospsyx What do you think? 🙂

xristos3490 replied:
I'd say to use capitalized letters since the "In" might not look great, but I'd wait for Mano's input! 

Another nit: It would be better to use the `_x()` function with context instead of the generic `__()`. These words are small, making it difficult to understand what is actually being translated.

manospsyx replied:
I'd probably:

- capitalize both (preferably via CSS; it's not great to rely on localizable strings for styling), and 
- use a slightly smaller font size, perhaps by 5-10%, or 1-2px :)

imanish003 replied:
@xristos3490 

> Another nit: It would be better to use the _x() function with context instead of the generic __(). These words are small, making it difficult to understand what is actually being translated.

Makes sense to me. What do you think about the following contexts?

```tsx
_x('Not in', 'Exclude products created within a certain time frame', 'woo-gutenberg-products-block');

_x('In', 'Include products created within a certain time frame', 'woo-gutenberg-products-block');
```

I can make the changes once we finalize the context text.

imanish003 replied:
@manospsyx 
> capitalize both (preferably via CSS; it's not great to rely on localizable strings for styling), and

Sounds good 👍 

> use a slightly smaller font size, perhaps by 5-10%, or 1-2px :)

I am utilizing the [ToggleGroupControl](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-togglegroupcontrol--docs) component from @wordpress/components. Given this, I am hesitant to manually adjust the font size with custom CSS since the component currently offers only two size variants: "default" and "__unstable-large." It may be prudent to retain the current sizing, and revisit this matter once more options are made available within the ToggleGroupControl. It's important to note that any future modifications should also be applied to the Layout toggle to maintain consistency between the two controls.

<img width="256" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/16707866/41e4c9a7-9493-4ecf-af48-01ab20f885b4">

What do you think? 🙂

manospsyx replied:
Good point, @imanish003 -- it's best to stick with the design of the component itself for consistency. In this case, let's go with capitalizing the first letter of the first word. 

On a related note and regarding the text string itself, I would use "In" / "Not In" to refer to lists of items = when the value is a multi-select input. 

In this case, I think it's more accurate/clear to go with:

- Within
- Before

imanish003 replied:
> let's go with capitalizing the first letter of the first word.

Sounds good to me 🙌🏻

> In this case, I think it's more accurate/clear to go with:
Within
Before

Makes sense. I will make the changes.

xristos3490 replied:
> Makes sense to me. What do you think about the following contexts?

I would keep this strictly to the broader context. Like:
```js
_x('Not in', 'Product Collection query operator', 'woo-gutenberg-products-block');
_x('In', 'Product Collection query operator', 'woo-gutenberg-products-block');
```

imanish003 replied:
@manospsyx I have made the changes & now UI looks like this:
<img width="281" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/16707866/c2623af7-1d19-428b-a7e5-0fe034374b08">

imanish003 replied:
@xristos3490 ~~Okay, I will go ahead & will make the changes. Thanks 🙏🏻~~

Maybe we don't need context anymore as we are now using "Within" & "Before" instead. What do you think?

xristos3490 replied:
It's a good practice for translators to have context when dealing with short phrases. There is no set rule for when to use the _x() function, but I believe this is a valid reason to use it.

That said, It's also advisable to provide context for other instances here like "Stack" or "Grid".

imanish003 replied:
@xristos3490 Makes sense. Let's try to figure out what text we can use for context 🙂

> _x('Not in', 'Product Collection query operator', 'woo-gutenberg-products-block');
_x('In', 'Product Collection query operator', 'woo-gutenberg-products-block');

This doesn't seem relevant anymore as we have changed the text to "Within" & "Before". What alternative text you would suggest? Here is my suggestion:
```
_x('Within', 'filter option for products created within a certain time frame', 'woo-gutenberg-products-block');
_x('Before', 'filter option for products created before a certain time', 'woo-gutenberg-products-block');
```

What do you think?

imanish003 replied:
> That said, It's also advisable to provide context for other instances here like "Stack" or "Grid".

Makes sense. What do you think about the following?
```
_x('Stack', 'layout style for displaying products in a single column', 'woo-gutenberg-products-block');
_x('Grid', 'layout style for displaying products in multiple columns', 'woo-gutenberg-products-block');
```

xristos3490 replied:
In my opinion, providing an exact description of the context could result in excessive wordiness. A simple term such as "Product Collection query operator" should suffice for translators to comprehend the intended meaning. However, I do not have a strong preference, so I leave it up to you to decide. :)

imanish003 replied:
I don't have a strong opinion either. I will go ahead and add "Product Collection query operator". Thanks 🙏🏻

Aljullu reviewed (approved):
This is testing well and the code looks good too. I left one minor comment, but approving on my behalf.

imanish003 committed:
Merge branch 'trunk' into add/11357-product-collection-filters-creation-date

imanish003 committed:
Merge branch 'trunk' into add/11357-product-collection-filters-creation-date

xristos3490 commented on src/BlockTypes/ProductCollection.php:
```diff
@@ -959,4 +964,38 @@ function( $rating ) use ( $product_visibility_terms ) {
+		return array(
+			'date_query' => array(
+				array(
+					'column'        => 'post_date',
```
Would it be more reliable to use `post_date_gmt` instead of `post_date` to account for DST?

imanish003 replied:
Good catch! Using post_date_gmt would indeed make our date queries DST-proof by relying on UTC. I'll update the PR to use post_date_gmt for better consistency across time zones. 🚀

imanish003 replied:
Made the changes in https://github.com/woocommerce/woocommerce-blocks/pull/11562/commits/b807d125859b115e53aaacf59d9adec87f26a5c9

xristos3490 reviewed (approved):
Nice work, @imanish003! 

I also left a few comments!

mikejolley commented:
Moving to 11.6.0

imanish003 committed:
Switch date query to use post_date_gmt for DST consistency

This commit changes the column reference in the date query from 'post_date' to 'post_date_gmt'. This update ensures that the product collection filtering is based on Coordinated Universal Time (UTC) rather than local time, which can be affected by Daylight Saving Time (DST) shifts. The modification will lead to more consistent and reliable behavior across different time zones and during DST changes.

imanish003 committed:
Capitalize toggle group labels

The following adjustments have been made:
- Introduced a constant `uppercaseStyle` to store the `{ textTransform: 'uppercase' }` style.
- Applied `uppercaseStyle` to both the 'IN' and 'NOT IN' toggle options to ensure label text is consistently uppercase.
- Updated the 'Not in' label text to uppercase ('NOT IN') to match the newly applied style.

These changes ensure that the toggle labels align with the design guidelines that call for uppercase styling in control elements.

imanish003 committed:
Merge branch 'trunk' into add/11357-product-collection-filters-creation-date

imanish003 committed:
Make first letter of first work capital

imanish003 committed:
Merge branch 'trunk' into add/11357-product-collection-filters-creation-date

imanish003 committed:
Rename to Within & Before

imanish003 committed:
Update i18n for Product Collection query operators

This commit updates the internationalization (i18n) for the Product Collection query operators in the 'Created' control component of the WooCommerce Blocks plugin. It replaces the '__' function with '_x' for translation and provides context comments for better translation handling. This improvement enhances the localization of the query operators for better multilingual support.

imanish003 committed:
Merge branch 'trunk' into add/11357-product-collection-filters-creation-date
